### PR TITLE
Demo/next translate

### PR DESCRIPTION
**Pull request by GautierArcin**
@timlrx  @rsipakov It should be finished now. I just added the sitemap. 

The branch doesn't seem to be able to merge automatically. [Maybe you can fetch it directly from my repo](https://github.com/GautierArcin/i18n-tailwind-nextjs-starter-blog/tree/demo/next-translate) ?  Branch is up-to-date with `timlrx/master`.

I'm not sure about three different things: 

1. I translated some [`sr-only` tags](https://tailwindcss.com/docs/screen-readers) with the `t` hook from next-translate. I'm not sure if they need to be purely statics (i.e. text) to work with screen readers ?

2. For tags, because tags are placed in the same directory, I had to localize them with the same system as articles ( `feed.xml` for default language and `feed.locale.xml` for others). I did not how to link them from the pages.

3. I would also be grateful if you could check the `sitemap` script. It seems to be working. I did a little of overthinking, though.  😅
I basically tried to put a system in place that detect automatically page presents in multiple language. The only caveat is that, for pages with multiples translation to work, they need to be at least the default language (which seems fine by me).

Otherwise, I'm  happy with the result. I'm also happy to contribute to this awesome template, and I hope my contribution will prove useful. 
Please let me know if you see things to change / things to improve !

(The branch might need some squash and merge though. 😅 )

### CODE REVIEW

**Comment by timlrx** [comment]
I think that it's really great, definitely far beyond my initial intent with a demo and a much more comprehensive example 😄.

Sitemap.xml seems to have some errors though. I get the following error on the deployed page:

image

**Comment by GautierArcin** [comment]
@timlrx Indeed ! There were some tags missing in the sitemap declaration.

See [https://stackoverflow.com/questions/36193346/validation-of-xml-sitemap-urlset-with-xhtmllink-inside-url-element](https://stackoverflow.com/questions/36193346/validation-of-xml-sitemap-urlset-with-xhtmllink-inside-url-element).

It should work normally now. 😃

**Comment by timlrx** [comment]
I don't think so? Here's what I am seeing for the sitemap.xml

image

File: scripts/generate-sitemap.js [hunk 1]
```diff
@@ -1,7 +1,10 @@
 const fs = require('fs')
 const globby = require('globby')
+const path = require('path')
 const prettier = require('prettier')
+const { cpuUsage } = require('process')
```
Comment (timlrx) [inline scripts/generate-sitemap.js:5]: This is probably not needed :)

File: scripts/generate-sitemap.js [hunk 2]
```diff
@@ -14,28 +17,90 @@ const siteMetadata = require('../data/siteMetadata')
...
+      xsi:schemaLocation="http://www.sitemaps.org/schemas/sitemap/0.9
+      http://www.sitemaps.org/schemas/sitemap/0.9/sitemap.xsd
+      http://www.w3.org/1999/xhtml
+      http://www.w3.org/2002/08/xhtml/xhtml1-strict.xsd">
+            ${pagesWithLoc
+              .map(([path, loc, alreadyPresent]) => {
+                // @todo: Can you check especially here ?
+                const route = path.includes('/index') ? path.replace('/index', '') : path
+                if (
+                  path.includes(`/404.js`) ||
```
Comment (timlrx) [inline scripts/generate-sitemap.js:71]: `path.includes(`/404`) || alreadyPresent` would do. The path would not contain `.js` so currently the 404 page is still in the sitemap

File: scripts/generate-sitemap.js [hunk 3]
```diff
@@ -14,28 +17,90 @@ const siteMetadata = require('../data/siteMetadata')
...
+                }</loc>
+                  ${routeMultiLang.map(
+                    ([xe, xloc]) =>
+                      `
+                               <xhtml:link 
+                               rel="alternate"
+                               hreflang="${xloc}"
+                               href="${siteMetadata.siteUrl}${xe}"/>
+                               `
+                  )}
```
Comment (timlrx) [inline scripts/generate-sitemap.js:101]: Perhaps change this line to`).join('')}`? The comma looks a little off in the XML file though it does not throw any error in the validators I tried.

**Review by timlrx — COMMENTED** [review]
This is some complicated sitemap logic but it looks to be working!

File: scripts/generate-sitemap.js (hunk 1 above)
Comment (GautierArcin) [reply scripts/generate-sitemap.js]: Yup ! (That was totally VsCode fault 😶 )

**Comment by GautierArcin** [comment]
@timlrx All problems should be fixed now.

(I've been thinking, maybe instead of using that branch as a demo, we could try to maintain it, in order for people to be able to start their blog with `i18n` and `next-translate` ? They'll just need to config `i18n.json` in root and change `locales` folder accordingly). 🙂

**Comment by timlrx** [comment]
Sorry, should have asked you to resolve the merge conflict. I thought it would just change on my branch, but in any case, everything looks good now and I will proceed to merge it!

I definitely see the benefits of doing so, but I have to weigh it against maintaining a third branch (there's currently master and typescript) and that the code is quite different. Getting everyone to use the version with `next-translate` also seems less than ideal, as modifying the code is quite a bit more opaque. I might try to incorporate the markdown locale sourcing part in the main codebase, then that leaves implementing the `next-translate` layer to the user - which is quite doable but tedious. I acknowledge sitemap is a pain though haha.

**Comment by GautierArcin** [comment]
No worries ! They were very minor merge conflcts. Thank you for the merge !

I totally understand your point of view.  (Yeah sitemap, was a totally a pain. I guess you can tell by how the sitemap is totally overthinked now...  😅 )

What I might do is, trying to maintain a branch (`next-translate` or `demo/next-translate` ?) on my fork of your repo with `master` of this repo.
I'm using that branch for my blog and would like to keep it up to date with your repo. If that can benefits other users, that's a win in my book. 
I might have to think a little bit more about it though (because it might implies that my blog, with custom ressources, and the `next-translate` branch might need to be in the same repo).

Also, I deeply apologize, but while you were doing the merge, I was checking the sources, and I'm not sure about two differents things. I might have forgot things for the SEO. Could you tell me what you think ? 

1. I generate a `feed.{locale}.xml` for each locale that is not the defaultLocale.  But the other locale are not present in `_document.js`. 

Shall we do a ` <link rel="alternate" type="application/rss+xml" href="/feed.xml" /> ` (l24 of `_document.js`) for every non-default locale(s) ?

2. Kind of the same question for `SEO.js` component.  
Shall we change  `<link rel="canonical" href={`${siteMetadata.siteUrl}${router.asPath}`} /> ` (l104) ? 
Thing is, to change it ""correctly"", we might to assume that every articles is present in every locales (`BlogSEO` is loaded from two layouts). And I'm not super comfortable with that assumption.

Sorry again for these two questions after the merge.

**Comment by timlrx** [comment]
Sure, I have already linked your fork on the readme!

Not the most knowledgeable on the i18n front, but it seems that the standard is to have a `site/{locale}/.../feed.xml` and specify the [element channel language](https://www.rssboard.org/rss-specification#element-channel-language) as part of the feed. Here's an example of [UBS media feed](https://www.ubs.com/global/en/media/notifications.html)

I don't see any clear guidelines on the RSS auto discovery (`<link rel="alternate".../>`), but it seems that multiple alternative feeds is fine.

For SEO, [i18n nuxtjs](https://i18n.nuxtjs.org/seo/) provides a good set of guidelines. Yes, it seems that the canonical link should point to the main page but it is an issue that it might not exist if for example the article for the default language is missing... One possibility (quite a different change in approach) would be to scan through all the files at the start and build some kind of dictionary that can populate the SEO and maybe also the sitemap later. Also, it seems that we should add a `og:locale` and `og:locale:alternate` meta tag.

Don't worry about the questions. I foresee a bit more struggle until we converge to a nice solution 😄

**Comment by GautierArcin** [comment]
I do agree on the `site/{locale}/.../feed.xml` idea. Problem is that Next-js seems to consider  `site/{locale}` as `site`, with just the `locale` props that change. 
Example: [english feed on french local](https://tailwind-nextjs-starter-blog-git-demo-next-cd29dc-gautierarcin.vercel.app/fr/feed.xml); [french feed on english local](https://tailwind-nextjs-starter-blog-git-demo-next-cd29dc-gautierarcin.vercel.app/feed.fr.xml) (fr/feed.xml and /feed.fr.xml). 
If we create a `fr` directory, for exemple, in `statics`, the new route of the xml would be `site/fr/fr/feed.xml`, for example...

I've been wrestling for two hours to try to get the {locale) working inside the `_document.js`, in order to add the auto-discovery of rss feed depending on the local... I wasn't able to get the `local` inside the `render()` function, which prevented me to change the auto-discovery.

I do agree on the SEO. I'm not familiar enough with next-js, how can you execute a script once before the build ? Modifying the `build` command in `package.json` like you did ?
I think we should take particular precaution to execute it only once. 
Or maybe we can the three ssg pages (`[page].js`, `[...slug].js` && `[tag].js` to check it while doing their SSG ? They would have to do other call to `getAllFiles...()`, which might be costly in terms of performance, but it might be way cleaner in my humble opinion (since it would just be another condition on an empty (or not) array in `SEO.js`).

(Thanks for the Readme ! `Gautier` is without an `h` though 😅).

**Comment by timlrx** [comment]
Will get back on `feed.xml`.

`_document.js` is rendered in the server so I don't think it would support the auto-discovery depending on locale. In fact, *all* `feed.xml` are listed regardless of locale.

To execute a script before build you should be able to just run `node ./script.js && next build`. Yup, modify the npm build script. Yes, if we are to do it that way then it makes sense to refer to this parsed file instead of having to call `getAllFiles` repeatedly. But, it is a big refactor 😬 

Fixed the typo, my bad!

**Comment by timlrx** [comment]
I think for a statically generated sitemap, your current approach seems like the best we can do. The other alternative is to generate the sitemap server side like this [tutorial](https://cheatcode.co/tutorials/how-to-generate-a-dynamic-sitemap-with-next-js) but this would also mean that that the template will no longer be 100% static.

Also, I take back my suggestion to run a prep script before running build. The main downside is that this will break a the hot reloading and affect user experience while developing, since to sync every change, the prep script would have to run first.

**Comment by GautierArcin** [comment]
I think I just finished. Let me know what you think. Suggested RSS feed and SEO should work. 😁


[Live preview](
tailwind-nextjs-starter-blog-git-demo-next-cd29dc-gautierarcin.vercel.app) & [Github branch](https://github.com/GautierArcin/i18n-tailwind-nextjs-starter-blog/commits/demo/next-translate).

Let me know what you think !

Also, I think you should look into [this commit](https://github.com/GautierArcin/i18n-tailwind-nextjs-starter-blog/commit/910f4c548390f4433eef6fe22fd5399ce7d39481). Rss suggestion on tags were not working.

I opened a PR on main repo ( #182 ).

**Comment by rsipakov** [comment]
> I think I just finished. Let me know what you think. Suggested RSS feed and SEO should work. 😁
> 
> [Live preview](tailwind-nextjs-starter-blog-git-demo-next-cd29dc-gautierarcin.vercel.app) & [Github branch](https://github.com/GautierArcin/i18n-tailwind-nextjs-starter-blog/commits/demo/next-translate).
> 
> Let me know what you think !
> 
> Also, I think you should look into [this commit](https://github.com/GautierArcin/i18n-tailwind-nextjs-starter-blog/commit/910f4c548390f4433eef6fe22fd5399ce7d39481). Rss suggestion on tags were not working.
> 
> I opened a PR on main repo ( #182 ).

@GautierArcin I seem to be way behind on this topic. Do I understand correctly, now we have feed.xml and feed.fr.xml? 

Also, for sitemap.xml I have like [this](https://tailwind-nextjs-starter-blog-rsipakov-bjm320fuv-jamstack-os.vercel.app/sitemap.xml):

```
https://tailwind-nextjs-starter-blog.vercel.app/about https://tailwind-nextjs-starter-blog.vercel.app/blog https://tailwind-nextjs-starter-blog.vercel.app/index https://tailwind-nextjs-starter-blog.vercel.app/projects https://tailwind-nextjs-starter-blog.vercel.app/tags https://tailwind-nextjs-starter-blog.vercel.app/blog/introducing-tailwind-nextjs-starter-blog https://tailwind-nextjs-starter-blog.vercel.app/blog/new-features-in-v1 https://tailwind-nextjs-starter-blog.vercel.app/blog/new-features-in-v2 https://tailwind-nextjs-starter-blog.vercel.app/tags/common-tag https://tailwind-nextjs-starter-blog.vercel.app/tags/guide https://tailwind-nextjs-starter-blog.vercel.app/tags/next-js https://tailwind-nextjs-starter-blog.vercel.app/fr/tags/tag-francais https://tailwind-nextjs-starter-blog.vercel.app/tags/tailwind
```

Is it correct or is this issue on my side? Thank you!

**Comment by timlrx** [comment]
@GautierArcin I think it works well. Nice job generating the SEO permutations 👍 

@rsipakov not sure why the feed displays like that on the browser. I use chrome and see the same thing, but it parses as a valid sitemap from the validators that I checked.

**Comment by rsipakov** [comment]
> @GautierArcin I think it works well. Nice job generating the SEO permutations 👍
> 
> @rsipakov not sure why the feed displays like that on the browser. I use chrome and see the same thing, but it parses as a valid sitemap from the validators that I checked.

@GautierArcin It is really nice job! 
@timlrx Thank you for this notice about the sitemap. But, some validators show errors with this sitemap. For example [inspyder.com](https://www.inspyder.com/free/SitemapValidator.aspx).

**Comment by GautierArcin** [comment]
@timlrx  @rsipakov  Thanks you a lot ! I think it's ready to merge. At least, I don't see anything to change right now.

@rsipakov  I think it shows like that because of `<loc>` and `url` tag. Every "" canonical"" url is shown. 
For the package, it could be interesting. Is the package only necessary at development (To not make the bundle size larger) ?

Also, problem is that, you can't "" check "" if url exist in local or not with this package.
What I mean is that `/script/generate-sitemap.js` is now quite complicated because we try to check if the current page is available in multiple locale. If so, we can precise in the sitemap that both link are the same content, but different langage (to not be doxxed by Google SEO). I fear it would be difficult to do so with this package. 😅

**Comment by vahidtakro** [comment]
That's awesome! good job!

this is how I managed to translate the Disqus comments, although it's not my favorite system and I'm in search for a simple one. 

```
import { useRouter } from 'next/router'
...
  const { t } = useTranslation()
  const { locale } = useRouter()
  const Language = locale === 'en' ? 'en' : 'fr'
...
    window.disqus_config = function () {
      this.language = `${Language}`
    }
```

**Comment by timlrx** [comment]
@GautierArcin sure, we can merge the latest one. Kindly create a new pull request, thanks!

**Comment by GautierArcin** [comment]
I don't understand your question. Can you reformulate ?

**Comment by vahidtakro** [comment]
Sure, if we create a js file in "Pages" folder, for example "music" , we can access that js page as mywebsite.com/music , with our js content, like putting a music playlist, etc. This URL will be accessible for both locales, and we need to translate the values with next-translate package. My question is, what if we want to have that page (js file) only in a specific language?

**Comment by rsipakov** [comment]
> Sure, if we create a js file in "Pages" folder, for example "music.js" , we can access that js page as mywebsite.com/music , with our js content, like putting a music playlist, etc. This URL will be accessible for both locales, and we need to translate the values with next-translate package. My question is, what if we want to have that page (js file) only in a specific language?

@vahidtakro as I understood your question and main concept for "next-translate" on firs step you need to study how is it work. Quick answer on your question: for long content you can use only md/mdx translation concept, for short content you can use hook, like (t).

**Comment by timlrx** [comment]
If I understand the question correctly, @vahidtakro is asking whether it is possible to make a certain page only for one particular language and not the other. Should be possible by specifying the routes you want in `getStaticPaths`. The pseudocode below should generate the page only for the `fr` route:

```
export const getStaticPaths = ({ locales }) => {
  return {
    paths: [
      { params: { }, locale: 'fr' },
    ],
  }
}
```

**Comment by rsipakov** [comment]
> If I understand the question correctly, @vahidtakro is asking whether it is possible to make a certain page only for one particular language and not the other. Should be possible by specifying the routes you want in `getStaticPaths`. The pseudocode below should generate the page only for the `fr` route:
> 
> 
> 
> ```
> 
> export const getStaticPaths = ({ locales }) => {
> 
>   return {
> 
>     paths: [
> 
>       { params: { }, locale: 'fr' },
> 
>     ],
> 
>   }
> 
> }
> 
> ```

 Very useful tip.

**Comment by GautierArcin** [comment]
@vahidtakro  I do apologize to not having respond earlier. Timothy gave you the right way of doing it.

@timlrx I just merged update from master of this main repo into `demo/next-translate` branch. It's now again up to date.

Also, I included a succinct Readme [on the repo](https://github.com/GautierArcin/i18n-tailwind-nextjs-starter-blog), if someone is interested to do a blog with multiple locales.

**Comment by vahidtakro** [comment]
Thanks @GautierArcin and @timlrx, I just don't know where to add it,

and also a question got my mind so busy, how can I use this localized template with [domain routing](https://nextjs.org/docs/advanced-features/i18n-routing#domain-routing)?

**Comment by GautierArcin** [comment]
> Thanks @GautierArcin and @timlrx, I just don't know where to add it,
> 
> and also a question got my mind so busy, how can I use this localized template with [domain routing](https://nextjs.org/docs/advanced-features/i18n-routing#domain-routing)?
> 
> having this would will redirect both domains to the main default domain, which is English.
> 
> # i18n.json
> ```
> {
>   "locales": ["en", "fr"],
>   "defaultLocale": "en",
>   "domains": [
>     {
>       "domain": "mydomain.com",
>       "defaultLocale": "en"
>     },
>     {
>       "domain": "myanotherdomain.fr",
>       "defaultLocale": "fr"
>     }
>   ],
>   "pages": {
>     "*": ["common", "headerNavLinks", "404"],
>     "/projects": ["projects"],
>   }
> }
> ```

You would need to add it to your 'music.js' page, in your example.

I honestly don't know for sub domain routing. My guess is that you'll have to set up the dns for both domains (that's why I used sub-path routing instead of dns routing). Otherwhise, I don't know.

In order for it to work effectively with the `i18n` template, you'll have to modify a lot of route (from `example.com/{locale}/article` to `example.{locale}/article`), especially for SEO component. I think that would prouve tedious and time consuming.

**Comment by vahidtakro** [comment]
> > Thanks @GautierArcin and @timlrx, I just don't know where to add it,
> > and also a question got my mind so busy, how can I use this localized template with [domain routing](https://nextjs.org/docs/advanced-features/i18n-routing#domain-routing)?
> > having this would will redirect both domains to the main default domain, which is English.
> > # i18n.json
> > ```
> > {
> >   "locales": ["en", "fr"],
> >   "defaultLocale": "en",
> >   "domains": [
> >     {
> >       "domain": "mydomain.com",
> >       "defaultLocale": "en"
> >     },
> >     {
> >       "domain": "myanotherdomain.fr",
> >       "defaultLocale": "fr"
> >     }
> >   ],
> >   "pages": {
> >     "*": ["common", "headerNavLinks", "404"],
> >     "/projects": ["projects"],
> >   }
> > }
> > ```
> 
> You would need to add it to your 'music.js' page, in your example.
> 
> I honestly don't know for sub domain routing. My guess is that you'll have to set up the dns for both domains (that's why I used sub-path routing instead of dns routing). Otherwhise, I don't know.
> 
> In order for it to work effectively with the `i18n` template, you'll have to modify a lot of route (from `example.com/{locale}/article` to `example.{locale}/article`), especially for SEO component. I think that would prouve tedious and time consuming.

thanks, I am still trying for two days, no result yet. I don't know what the problem is , it's about the template, or about the changes for internationalizing, or vercel or a nextjs bug, 

I expected to switch to my locales like this:
domain.com for English,
domain.fr for French,

but I'm experiencing this:

domain.com for English
domain.fr for English

domain.com/fr for French
domain.fr/fr for French

**Comment by GautierArcin** [comment]
Let me preface this by saying that I might be wrong. I did not try domain routing and don't intend to.

According to the documentation I read, you have to setup the `locales` and  `defaultLocale` when setting up domain routing. Thing is, when you setup both these values, you setup sub-path routing.

Judging by your problem, it seems that Vercel (or next) recognize both domains as part of your site, but still use sub-path routing on both domains.

Therefore, I would advise you to look at your `domains` array in your `i18n` config. I would also advise you to look at your DNS/domains settings in your deployments. 

Sorry to not help more.

**Comment by GautierArcin** [comment]
No worries, glad I could help !

**Comment by GautierArcin** [comment]
No worries, glad I could help !

**Comment by vahidtakro** [comment]
Hello, today I figured out something new, I just posted something in my second language, so only made .fr.md file for it, and now I saw that when I go in that post (in deployed app) I see an error which the app is still looking for the en version but it's missing.

so this is the file which responds 404:
https://www.mywebsite.com/_next/data/7i3RTEobxVMq6P_OeBv92/en/diary/site-has-changed.json

Capkture

**Comment by GautierArcin** [comment]
I was not able to replicate the bug on the starter. Once again, please provide a branch on the `i18n` starter if you want me to look into it.

**Pull request merged**